### PR TITLE
spark agent #571 fix dto mapper

### LINKED ISSUE #571: ModelMapper in some cases doesn't map AttrOrExprRef to proper api class
Happens when mapping from internal agent model into the producer api model.

The internal model class stays unmapped an is converted to json that causes issue further down the line.

### PR DESCRIPTION
**Pull request by cerveada**
fixes #571

### CODE REVIEW

File: core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala [hunk 1]
```diff
@@ -16,48 +16,110 @@
...
+    extra = toUntypedMap(attribute.extra).asNonEmptyOption,
+    name = attribute.name
+  )
+
+  def toAttrOrExprRef(o: AttrOrExprRef): v1_2.AttrOrExprRef = o match {
     case AttrRef(attrId) => v1_2.AttrOrExprRef(Some(attrId), None)
     case ExprRef(exprId) => v1_2.AttrOrExprRef(None, Some(exprId))
   }
+
+  val untypedMapConverter = new UntypedMapConverter(toAttrOrExprRef)
```
Comment (wajda) [inline core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala]: Why do you need a shared instance of this class? From what I see it doesn't have any state, so can be just a function or an object, no?

File: core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala [hunk 2]
```diff
@@ -16,48 +16,110 @@
...
+  def toConstant(l: Literal): v1_2.Literal = v1_2.Literal(
+    id = l.id,
+    dataType = l.dataType,
+    extra = toUntypedMap(l.extra).asNonEmptyOption,
+    value = l.value
+  )
+
+  def toNameAndVersion(nav: NameAndVersion): v1_2.NameAndVersion = v1_2.NameAndVersion(
+    name = nav.name, version = nav.version
+  )
```
Comment (wajda) [inline core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala]: formatting: either write the entire call statement in one line, or place one arg per line.

File: core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala [hunk 3]
```diff
@@ -16,23 +16,19 @@
...
     }
-  }
 
-  implicit def mapToOption[K, V](map: Map[K, V]): Option[Map[K, V]] =
-    if (map.isEmpty) {
-      None
-    } else {
-      Some(map)
-    }
+    map.transform((k, v) => mapValue(v))
```
Comment (wajda) [inline core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala]: `k` is unused. Wouldn't  `map.mapValue()` be enough? (same above)

Comment (cerveada) [reply core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala]: I don't like how it hides exceptions https://dev.to/cerveada/scala-tip-mapvalues-may-not-do-what-you-expect-2n2e

Comment (wajda) [reply core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala]: hide? Do you mean throw it on a later stage?

Comment (cerveada) [reply core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala]: yes

File: core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala (hunk 1 above)
Comment (cerveada) [reply core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala]: I changed it to function

File: core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala (hunk 3 above)
Comment (wajda) [reply core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/UntypedMapConverter.scala]: Got it. Just a minor suggestion for future - replace unused arguments with `_` for easier reading.

**Review by wajda — APPROVED** [review]
LGTM

**Pull request merged**